#📌 Extracción

In [108]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [109]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [110]:
df = pd.read_json('/content/drive/MyDrive/Colab Notebooks/ESTADISTICA_Y_MACHINE_LEARNIG/CHALLENGE/TelecomX_Data.json')
df


customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7267 rows x 6 columns]

#🔧 Transformación

In [111]:
# Normalizar cada una de las columnas del dataframe
df_customer = pd.json_normalize(df['customer'])
df_phone = pd.json_normalize(df['phone'])
df_internet = pd.json_normalize(df['internet'])
df_account = pd.json_normalize(df['account'])

In [112]:
# Extrae las columnas que no son JSON (customerID y Churn)
customerID = df['customerID']
Churn = df['Churn']


In [113]:
# concatenar las columnas en un solo dataframe
df = pd.concat([customerID, Churn, df_customer, df_phone, df_internet, df_account], axis=1)

df.sample(5)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
6296  8659-HDIYE    No  Female              1      No         No      64   
2167  3030-YZADT   Yes    Male              0      No         No       9   
7257  9975-SKRNR    No    Male              0      No         No       1   
5038  6892-EZDTG   Yes  Female              0     Yes         No       4   
1954  2774-LVQUS   Yes  Female              1     Yes         No      15   

     PhoneService MultipleLines InternetService  ...         OnlineBackup  \
6296          Yes           Yes             DSL  ...                  Yes   
2167          Yes            No             DSL  ...                   No   
7257          Yes            No              No  ...  No internet service   
5038          Yes           Yes     Fiber optic  ...                  Yes   
1954          Yes           Yes     Fiber optic  ...                   No   

         DeviceProtection          TechSupport          StreamingTV  \
6296                  Yes                  Yes                   No   
2167                   No                   No                   No   
7257  No internet service  No internet service  No internet service   
5038                   No                   No                  Yes   
1954                  Yes                  Yes                   No   

          StreamingMovies        Contract PaperlessBilling  \
6296                  Yes  Month-to-month               No   
2167                   No  Month-to-month               No   
7257  No internet service  Month-to-month               No   
5038                   No  Month-to-month              Yes   
1954                   No  Month-to-month              Yes   

                PaymentMethod Charges.Monthly  Charges.Total  
6296  Credit card (automatic)           74.65        4869.35  
2167             Mailed check           44.95            431  
7257             Mailed check           18.90           18.9  
5038         Electronic check           91.65          365.4  
1954         Electronic check           83.05         1258.3  

[5 rows x 21 columns]

In [114]:
# Abrir archivo .md y mostrar el contenido del diccionario como texto
with open('/content/drive/MyDrive/Colab Notebooks/ESTADISTICA_Y_MACHINE_LEARNIG/CHALLENGE/TelecomX_diccionario.md', 'r', encoding='utf-8') as f:
    contenido = f.read()

print(type(contenido))  # <class 'str'>
print(contenido)        # Imprime el contenido como texto


<class 'str'>
#### Diccionario de datos

- `customerID`: número de identificación único de cada cliente
- `Churn`: si el cliente dejó o no la empresa
- `gender`: género (masculino y femenino)
- `SeniorCitizen`: información sobre si un cliente tiene o no una edad igual o mayor a 65 años
- `Partner`: si el cliente tiene o no una pareja
- `Dependents`: si el cliente tiene o no dependientes
- `tenure`: meses de contrato del cliente
- `PhoneService`: suscripción al servicio telefónico
- `MultipleLines`: suscripción a más de una línea telefónica
- `InternetService`: suscripción a un proveedor de internet
- `OnlineSecurity`: suscripción adicional de seguridad en línea
- `OnlineBackup`: suscripción adicional de respaldo en línea
- `DeviceProtection`: suscripción adicional de protección del dispositivo
- `TechSupport`: suscripción adicional de soporte técnico, menor tiempo de espera
- `StreamingTV`: suscripción de televisión por cable
- `StreamingMovies`: suscripción de streaming de películas
-

In [115]:
# Ver cuántos valores nulos hay por columna
df.isnull().sum()


customerID          0
Churn               0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges.Monthly     0
Charges.Total       0
dtype: int64

In [116]:
# Mostrar todas las filas que tienen al menos un valor nulo
print ('El numero de valores nulos es:', len(df[df.isnull().any(axis=1)]))

El numero de valores nulos es: 0


In [117]:
df.shape

(7267, 21)

In [118]:
# Filas que están duplicadas exactamente (incluida la primera aparición)
duplicados_completos = df[df.duplicated(keep=False)]

# Opcional: Ordenar para que las filas duplicadas queden juntas
duplicados_ordenados = duplicados_completos.sort_values(by=df.columns.tolist())


print('El numero de filas duplicadas es: ', len(duplicados_ordenados))


El numero de filas duplicadas es:  0


In [119]:
df.head(1)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   

  DeviceProtection TechSupport StreamingTV StreamingMovies  Contract  \
0               No         Yes         Yes              No  One year   

  PaperlessBilling PaymentMethod Charges.Monthly  Charges.Total  
0              Yes  Mailed check            65.6          593.3  

[1 rows x 21 columns]

In [120]:
# Verificar los tipos de datos
df.dtypes

customerID           object
Churn                object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
Charges.Monthly     float64
Charges.Total        object
dtype: object

In [121]:
# Verificacion del tipo de dato de cada una una de las celdas de las columnas
columnas = [
    'customerID',	'Churn', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines',
    'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
    'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
    'PaperlessBilling', 'PaymentMethod', 'Charges.Monthly', 'Charges.Total'
]

for col in columnas:
    print(f"Tipos en la columna '{col}':")
    print(df[col].apply(type).value_counts())
    print(df[col].isna().sum())
    print("-" * 40)


Tipos en la columna 'customerID':
customerID
<class 'str'>    7267
Name: count, dtype: int64
0
----------------------------------------
Tipos en la columna 'Churn':
Churn
<class 'str'>    7267
Name: count, dtype: int64
0
----------------------------------------
Tipos en la columna 'gender':
gender
<class 'str'>    7267
Name: count, dtype: int64
0
----------------------------------------
Tipos en la columna 'SeniorCitizen':
SeniorCitizen
<class 'int'>    7267
Name: count, dtype: int64
0
----------------------------------------
Tipos en la columna 'Partner':
Partner
<class 'str'>    7267
Name: count, dtype: int64
0
----------------------------------------
Tipos en la columna 'Dependents':
Dependents
<class 'str'>    7267
Name: count, dtype: int64
0
----------------------------------------
Tipos en la columna 'tenure':
tenure
<class 'int'>    7267
Name: count, dtype: int64
0
----------------------------------------
Tipos en la columna 'PhoneService':
PhoneService
<class 'str'>    7267
Nam

In [122]:
# Cambio de tipo 'object' de la columna 'Charges.Total' a tipo 'float'
df['Charges.Total'] = pd.to_numeric(df['Charges.Total'], errors='coerce')

# Verifica si la conversión funcionó
print(df['Charges.Total'].apply(type).value_counts())
print(df['Charges.Total'].isna().sum())



Charges.Total
<class 'float'>    7267
Name: count, dtype: int64
11


In [123]:
#Existen 11 valores nulos
df_nulos = df[df['Charges.Total'].isna()]
display(df_nulos)



customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
975   1371-DWPAZ    No  Female              0     Yes        Yes       0   
1775  2520-SGTTA    No  Female              0     Yes        Yes       0   
1955  2775-SEFEE    No    Male              0      No        Yes       0   
2075  2923-ARZLG    No    Male              0     Yes        Yes       0   
2232  3115-CZMZD    No    Male              0      No        Yes       0   
2308  3213-VVOLG    No    Male              0     Yes        Yes       0   
2930  4075-WKNIU    No  Female              0     Yes        Yes       0   
3134  4367-NUYAO    No    Male              0     Yes        Yes       0   
3203  4472-LVYGI    No  Female              0     Yes        Yes       0   
4169  5709-LVOEQ    No  Female              0     Yes        Yes       0   
5599  7644-OMVMY    No    Male              0     Yes        Yes       0   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
975            No  No phone service             DSL  ...                  Yes   
1775          Yes                No              No  ...  No internet service   
1955          Yes               Yes             DSL  ...                  Yes   
2075          Yes                No              No  ...  No internet service   
2232          Yes                No              No  ...  No internet service   
2308          Yes               Yes              No  ...  No internet service   
2930          Yes               Yes             DSL  ...                  Yes   
3134          Yes               Yes              No  ...  No internet service   
3203           No  No phone service             DSL  ...                   No   
4169          Yes                No             DSL  ...                  Yes   
5599          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
975                   Yes                  Yes                  Yes   
1775  No internet service  No internet service  No internet service   
1955                   No                  Yes                   No   
2075  No internet service  No internet service  No internet service   
2232  No internet service  No internet service  No internet service   
2308  No internet service  No internet service  No internet service   
2930                  Yes                  Yes                  Yes   
3134  No internet service  No internet service  No internet service   
3203                  Yes                  Yes                  Yes   
4169                  Yes                   No                  Yes   
5599  No internet service  No internet service  No internet service   

          StreamingMovies  Contract PaperlessBilling  \
975                    No  Two year               No   
1775  No internet service  Two year               No   
1955                   No  Two year              Yes   
2075  No internet service  One year              Yes   
2232  No internet service  Two year               No   
2308  No internet service  Two year               No   
2930                   No  Two year               No   
3134  No internet service  Two year               No   
3203                   No  Two year              Yes   
4169                  Yes  Two year               No   
5599  No internet service  Two year               No   

                  PaymentMethod Charges.Monthly  Charges.Total  
975     Credit card (automatic)           56.05            NaN  
1775               Mailed check           20.00            NaN  
1955  Bank transfer (automatic)           61.90            NaN  
2075               Mailed check           19.70            NaN  
2232               Mailed check           20.25            NaN  
2308               Mailed check           25.35            NaN  
2930               Mailed check           73.35            NaN  
3134               Mailed check           25.75            NaN  
3203  Ban

In [124]:
#Para cada fila donde 'Charges.Total' es NaN, reemplazar ese NaN con el valor de 'Charges.Monthly' * 0.5'
#ya que la columna tenure o antiguedad en meses es igual a cero en estos casos debido a que es primer mes del cliente
#por lo que se va usar 'Charges.Monthly' * 0.5' como valor promedio para 'Charges.Total'

# La formulacion es: df.loc[condición, 'columna'] = nuevo_valor
df.loc[df['Charges.Total'].isna(), 'Charges.Total'] = (df['Charges.Monthly'] * 0.5)


In [125]:
#Verificacion de transformacion de valores NaN
df_nulos = df[df['Charges.Total'].isna()]
display(df_nulos)

Empty DataFrame
Columns: [customerID, Churn, gender, SeniorCitizen, Partner, Dependents, tenure, PhoneService, MultipleLines, InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies, Contract, PaperlessBilling, PaymentMethod, Charges.Monthly, Charges.Total]
Index: []

[0 rows x 21 columns]

In [126]:
# verificar si fueron cambiados correctamente
df.iloc[[975, 1955, 2930, 5599]]



customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
975   1371-DWPAZ    No  Female              0     Yes        Yes       0   
1955  2775-SEFEE    No    Male              0      No        Yes       0   
2930  4075-WKNIU    No  Female              0     Yes        Yes       0   
5599  7644-OMVMY    No    Male              0     Yes        Yes       0   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
975            No  No phone service             DSL  ...                  Yes   
1955          Yes               Yes             DSL  ...                  Yes   
2930          Yes               Yes             DSL  ...                  Yes   
5599          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
975                   Yes                  Yes                  Yes   
1955                   No                  Yes                   No   
2930                  Yes                  Yes                  Yes   
5599  No internet service  No internet service  No internet service   

          StreamingMovies  Contract PaperlessBilling  \
975                    No  Two year               No   
1955                   No  Two year              Yes   
2930                   No  Two year               No   
5599  No internet service  Two year               No   

                  PaymentMethod Charges.Monthly  Charges.Total  
975     Credit card (automatic)           56.05         28.025  
1955  Bank transfer (automatic)           61.90         30.950  
2930               Mailed check           73.35         36.675  
5599               Mailed check           19.85          9.925  

[4 rows x 21 columns]

In [127]:
#verificar que los demas registros no fuero cambiados
df.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6         593.30  
1               No      Mailed check            59.9         542.40  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9         267.40  

[5 rows x 21 columns]

In [128]:
# Ultima verificación de columna Charges.Total
print(df['Charges.Total'].apply(type).value_counts())
print(df['Charges.Total'].isna().sum())

Charges.Total
<class 'float'>    7267
Name: count, dtype: int64
0


In [129]:
# Estandarización de los encabezados de las columnas
import re

# Función para estandarizar los nombres de las columnas
def estandarizar_columna(nombre):
    nombre = re.sub(r'(?<=[a-z])(?=[A-Z])', '_', nombre)  # separa camel case, Reemplazando cada coincidencia del patrón con un guion bajo (_) en la variable nombre.
    nombre = nombre.replace('.', '_')                     # reemplaza puntos
    nombre = nombre.strip().lower()                       # minúsculas y sin espacios
    return nombre

# Aplicar la función a los nombres de las columnas
df.columns = [estandarizar_columna(col) for col in df.columns]
df.columns


Index(['customer_id', 'churn', 'gender', 'senior_citizen', 'partner',
       'dependents', 'tenure', 'phone_service', 'multiple_lines',
       'internet_service', 'online_security', 'online_backup',
       'device_protection', 'tech_support', 'streaming_tv', 'streaming_movies',
       'contract', 'paperless_billing', 'payment_method', 'charges_monthly',
       'charges_total'],
      dtype='object')

In [130]:
df.head()

customer_id churn  gender  senior_citizen partner dependents  tenure  \
0  0002-ORFBO    No  Female               0     Yes        Yes       9   
1  0003-MKNFE    No    Male               0      No         No       9   
2  0004-TLHLJ   Yes    Male               0      No         No       4   
3  0011-IGKFF   Yes    Male               1     Yes         No      13   
4  0013-EXCHZ   Yes  Female               1     Yes         No       3   

  phone_service multiple_lines internet_service  ... online_backup  \
0           Yes             No              DSL  ...           Yes   
1           Yes            Yes              DSL  ...            No   
2           Yes             No      Fiber optic  ...            No   
3           Yes             No      Fiber optic  ...           Yes   
4           Yes             No      Fiber optic  ...            No   

  device_protection tech_support streaming_tv streaming_movies  \
0                No          Yes          Yes               No   
1                No           No           No              Yes   
2               Yes           No           No               No   
3               Yes           No          Yes              Yes   
4                No          Yes          Yes               No   

         contract paperless_billing    payment_method charges_monthly  \
0        One year               Yes      Mailed check            65.6   
1  Month-to-month                No      Mailed check            59.9   
2  Month-to-month               Yes  Electronic check            73.9   
3  Month-to-month               Yes  Electronic check            98.0   
4  Month-to-month               Yes      Mailed check            83.9   

   charges_total  
0         593.30  
1         542.40  
2         280.85  
3        1237.85  
4         267.40  

[5 rows x 21 columns]

In [131]:
# Estandarización en los registros de las columnas 'object'

for col in df.select_dtypes(include='object').columns:
    df[col] = (
        df[col]
        .astype(str)                                 # Asegura strings
        .str.strip()                                 # Elimina espacios al inicio/final
        .str.lower()                                 # Convierte a minúsculas
        .str.replace(r'[\s\-]+', '_', regex=True)    # Reemplaza espacios y guiones medio por guion bajo
    )

df


customer_id churn  gender  senior_citizen partner dependents  tenure  \
0     0002_orfbo    no  female               0     yes        yes       9   
1     0003_mknfe    no    male               0      no         no       9   
2     0004_tlhlj   yes    male               0      no         no       4   
3     0011_igkff   yes    male               1     yes         no      13   
4     0013_exchz   yes  female               1     yes         no       3   
...          ...   ...     ...             ...     ...        ...     ...   
7262  9987_lutyd    no  female               0      no         no      13   
7263  9992_rramn   yes    male               0     yes         no      22   
7264  9992_ujoel    no    male               0      no         no       2   
7265  9993_lhieb    no    male               0     yes        yes      67   
7266  9995_hotoh    no    male               0     yes        yes      63   

     phone_service    multiple_lines internet_service  ... online_backup  \
0              yes                no              dsl  ...           yes   
1              yes               yes              dsl  ...            no   
2              yes                no      fiber_optic  ...            no   
3              yes                no      fiber_optic  ...           yes   
4              yes                no      fiber_optic  ...            no   
...            ...               ...              ...  ...           ...   
7262           yes                no              dsl  ...            no   
7263           yes               yes      fiber_optic  ...            no   
7264           yes                no              dsl  ...           yes   
7265           yes                no              dsl  ...            no   
7266            no  no_phone_service              dsl  ...           yes   

     device_protection tech_support streaming_tv streaming_movies  \
0                   no          yes          yes               no   
1                   no           no           no              yes   
2                  yes           no           no               no   
3                  yes           no          yes              yes   
4                   no          yes          yes               no   
...                ...          ...          ...              ...   
7262                no          yes           no               no   
7263                no           no           no              yes   
7264                no           no           no               no   
7265               yes          yes           no              yes   
7266               yes           no          yes              yes   

            contract paperless_billing    payment_method charges_monthly  \
0           one_year               yes      mailed_check           65.60   
1     month_to_month                no      mailed_check           59.90   
2     month_to_month               yes  electronic_check           73.90   
3     month_to_month               yes  electronic_check           98.00   
4     month_to_month               yes      mailed_check           83.90   
...              ...               ...               ...             ...   
7262        one_year                no      mailed_check           55.15   
7263  month_to_month               yes  electronic_check           85.10   
7264  month_to_month               yes      mailed_check           50.30   
7265        two_year                no      mailed_check           67.85   
7266        two_year                no  electronic_check           59.00   

      charges_total  
0            593.30  
1            542.40  
2            280.85  
3           1237.85  
4            267.40  
...             ...  
7262         742.90  
7263        1873.70  
7264          92.75  
7265        4627.65  
7266        3707.60  

[7267 rows x 21 columns]

In [132]:
#Mostrar valores únicos por cada columna
for col in df.columns:
    print(f"\nColumna: '{col}'")
    print(df[col].unique())



Columna: 'customer_id'
['0002_orfbo' '0003_mknfe' '0004_tlhlj' ... '9992_ujoel' '9993_lhieb'
 '9995_hotoh']

Columna: 'churn'
['no' 'yes' '']

Columna: 'gender'
['female' 'male']

Columna: 'senior_citizen'
[0 1]

Columna: 'partner'
['yes' 'no']

Columna: 'dependents'
['yes' 'no']

Columna: 'tenure'
[ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 11 37 49 66
 67 20 43 59 12 27  2 25 29 14 35 64 39 40  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 51 31 36 17 18 38 42
  0]

Columna: 'phone_service'
['yes' 'no']

Columna: 'multiple_lines'
['no' 'yes' 'no_phone_service']

Columna: 'internet_service'
['dsl' 'fiber_optic' 'no']

Columna: 'online_security'
['no' 'yes' 'no_internet_service']

Columna: 'online_backup'
['yes' 'no' 'no_internet_service']

Columna: 'device_protection'
['no' 'yes' 'no_internet_service']

Columna: 'tech_support'
['yes' 'no' 'no_internet_service']

Columna: 'streaming_tv'
['yes' 'no' 'no_internet_service']

Columna: 's

In [133]:
# Cantidad de cadenas vacias en la columna 'churn'

num_cadenas_vacias = (df['churn'] == '').sum()
print(num_cadenas_vacias)


224


Debido a que entre los datos que maneja la columna 'churn' esta el de cadena vacia:['No' 'Yes' ''], no es posible saber si esos clientes fueron dados de baja o no, por lo tando estas filas seran eliminadas y guardadas en un archivo, "clientes_por_vericar.csv", con el fin de que se defina la condición de sus clientes para hacer un analisis posterior. Ademas se puede verifica que la cantidad de filas eliminadas es 224 lo cual representa apenas: (224/7267) * 100 = 3.1%, lo cual no es representativo para nuestro analisis estadistico

In [134]:
# Guardar filas eliminadas

# Filtrar las filas donde 'churn' es una cadena vacía
filas_vacias = df[df['churn'] == '']

# Guardar esas filas en un archivo CSV
filas_vacias.to_csv('clientes_por_vericar.csv', index=False)



In [135]:
# verificacion del archivo 'clientes_por_vericar.csv'

df_clientes_por_vericar = pd.read_csv('/content/clientes_por_vericar.csv', keep_default_na=False)
df_clientes_por_vericar.head()


customer_id churn  gender  senior_citizen partner dependents  tenure  \
0  0047_zhdtw        female               0      no         no      11   
1  0120_yzlqa          male               0      no         no      71   
2  0154_qyhju          male               0      no         no      29   
3  0162_rzgmz        female               1      no         no       5   
4  0274_vvqoq          male               1     yes         no      65   

  phone_service multiple_lines internet_service  ...        online_backup  \
0           yes            yes      fiber_optic  ...                   no   
1           yes             no               no  ...  no_internet_service   
2           yes             no              dsl  ...                  yes   
3           yes             no              dsl  ...                  yes   
4           yes            yes      fiber_optic  ...                  yes   

     device_protection         tech_support         streaming_tv  \
0                   no                   no                   no   
1  no_internet_service  no_internet_service  no_internet_service   
2                   no                  yes                   no   
3                   no                  yes                   no   
4                  yes                   no                  yes   

      streaming_movies        contract paperless_billing  \
0                   no  month_to_month               yes   
1  no_internet_service        two_year               yes   
2                   no        one_year               yes   
3                   no  month_to_month                no   
4                  yes        one_year               yes   

              payment_method charges_monthly  charges_total  
0  bank_transfer_(automatic)           79.00         929.30  
1    credit_card_(automatic)           19.90        1355.10  
2           electronic_check           58.75        1696.20  
3    credit_card_(automatic)           59.90         287.85  
4  bank_transfer_(automatic)          103.15        6792.45  

[5 rows x 21 columns]

In [136]:
# Eliminar en el dataframe original, df, las filas donde la columna 'churn', tenga cadenas vacias

# Mantener solo las filas donde churn NO es cadena vacía
# Convertir en una copia explícita el resultado del filtro, porque sino se hace pandas podria
# pensar que la asignacion, df, es una vista, y va ha generar una Advertencia cuando adicione la columna 'charges_monthly'
# El df original ya no existe, a menos que lo hubieras guardado previamente en otra variable
# la variable df ahora apunta a ese nuevo DataFrame filtrado.
df = df[df['churn'] != ''].copy()
df


customer_id churn  gender  senior_citizen partner dependents  tenure  \
0     0002_orfbo    no  female               0     yes        yes       9   
1     0003_mknfe    no    male               0      no         no       9   
2     0004_tlhlj   yes    male               0      no         no       4   
3     0011_igkff   yes    male               1     yes         no      13   
4     0013_exchz   yes  female               1     yes         no       3   
...          ...   ...     ...             ...     ...        ...     ...   
7262  9987_lutyd    no  female               0      no         no      13   
7263  9992_rramn   yes    male               0     yes         no      22   
7264  9992_ujoel    no    male               0      no         no       2   
7265  9993_lhieb    no    male               0     yes        yes      67   
7266  9995_hotoh    no    male               0     yes        yes      63   

     phone_service    multiple_lines internet_service  ... online_backup  \
0              yes                no              dsl  ...           yes   
1              yes               yes              dsl  ...            no   
2              yes                no      fiber_optic  ...            no   
3              yes                no      fiber_optic  ...           yes   
4              yes                no      fiber_optic  ...            no   
...            ...               ...              ...  ...           ...   
7262           yes                no              dsl  ...            no   
7263           yes               yes      fiber_optic  ...            no   
7264           yes                no              dsl  ...           yes   
7265           yes                no              dsl  ...            no   
7266            no  no_phone_service              dsl  ...           yes   

     device_protection tech_support streaming_tv streaming_movies  \
0                   no          yes          yes               no   
1                   no           no           no              yes   
2                  yes           no           no               no   
3                  yes           no          yes              yes   
4                   no          yes          yes               no   
...                ...          ...          ...              ...   
7262                no          yes           no               no   
7263                no           no           no              yes   
7264                no           no           no               no   
7265               yes          yes           no              yes   
7266               yes           no          yes              yes   

            contract paperless_billing    payment_method charges_monthly  \
0           one_year               yes      mailed_check           65.60   
1     month_to_month                no      mailed_check           59.90   
2     month_to_month               yes  electronic_check           73.90   
3     month_to_month               yes  electronic_check           98.00   
4     month_to_month               yes      mailed_check           83.90   
...              ...               ...               ...             ...   
7262        one_year                no      mailed_check           55.15   
7263  month_to_month               yes  electronic_check           85.10   
7264  month_to_month               yes      mailed_check           50.30   
7265        two_year                no      mailed_check           67.85   
7266        two_year                no  electronic_check           59.00   

      charges_total  
0            593.30  
1            542.40  
2            280.85  
3           1237.85  
4            267.40  
...             ...  
7262         742.90  
7263        1873.70  
7264          92.75  
7265        4627.65  
7266        3707.60  

[7043 rows x 21 columns]

In [137]:
# verificar que se hayan eliminado las filas con cadena vacia de la columna 'churn'
print(f"Valores unicos en la columna 'churn': {df['churn'].unique()}")



Valores unicos en la columna 'churn': ['no' 'yes']


In [138]:
# Veificacion de numero de filas
df.shape

(7043, 21)

In [139]:
# "Restablecer índices"
df.reset_index(drop=True, inplace=True)


In [140]:
# Creacion de columna cuentas diarias
df['daily_account'] = df['charges_monthly'] / 30
df.head()

customer_id churn  gender  senior_citizen partner dependents  tenure  \
0  0002_orfbo    no  female               0     yes        yes       9   
1  0003_mknfe    no    male               0      no         no       9   
2  0004_tlhlj   yes    male               0      no         no       4   
3  0011_igkff   yes    male               1     yes         no      13   
4  0013_exchz   yes  female               1     yes         no       3   

  phone_service multiple_lines internet_service  ... device_protection  \
0           yes             no              dsl  ...                no   
1           yes            yes              dsl  ...                no   
2           yes             no      fiber_optic  ...               yes   
3           yes             no      fiber_optic  ...               yes   
4           yes             no      fiber_optic  ...                no   

  tech_support streaming_tv streaming_movies        contract  \
0          yes          yes               no        one_year   
1           no           no              yes  month_to_month   
2           no           no               no  month_to_month   
3           no          yes              yes  month_to_month   
4          yes          yes               no  month_to_month   

  paperless_billing    payment_method charges_monthly charges_total  \
0               yes      mailed_check            65.6        593.30   
1                no      mailed_check            59.9        542.40   
2               yes  electronic_check            73.9        280.85   
3               yes  electronic_check            98.0       1237.85   
4               yes      mailed_check            83.9        267.40   

   daily_account  
0       2.186667  
1       1.996667  
2       2.463333  
3       3.266667  
4       2.796667  

[5 rows x 22 columns]

In [141]:
# Reubicar la columna daily_account
# Paso 1: Sacar la columna
columna = df.pop('daily_account')

# Paso 2: Insertarla en la posición deseada (posición 20 → índice 19)
df.insert(19, 'daily_account', columna)
df.head()

customer_id churn  gender  senior_citizen partner dependents  tenure  \
0  0002_orfbo    no  female               0     yes        yes       9   
1  0003_mknfe    no    male               0      no         no       9   
2  0004_tlhlj   yes    male               0      no         no       4   
3  0011_igkff   yes    male               1     yes         no      13   
4  0013_exchz   yes  female               1     yes         no       3   

  phone_service multiple_lines internet_service  ... device_protection  \
0           yes             no              dsl  ...                no   
1           yes            yes              dsl  ...                no   
2           yes             no      fiber_optic  ...               yes   
3           yes             no      fiber_optic  ...               yes   
4           yes             no      fiber_optic  ...                no   

  tech_support streaming_tv streaming_movies        contract  \
0          yes          yes               no        one_year   
1           no           no              yes  month_to_month   
2           no           no               no  month_to_month   
3           no          yes              yes  month_to_month   
4          yes          yes               no  month_to_month   

  paperless_billing    payment_method daily_account charges_monthly  \
0               yes      mailed_check      2.186667            65.6   
1                no      mailed_check      1.996667            59.9   
2               yes  electronic_check      2.463333            73.9   
3               yes  electronic_check      3.266667            98.0   
4               yes      mailed_check      2.796667            83.9   

   charges_total  
0         593.30  
1         542.40  
2         280.85  
3        1237.85  
4         267.40  

[5 rows x 22 columns]

In [142]:
# redondear decimales a un digito
columnas = ['daily_account', 'charges_monthly', 'charges_total']
df[columnas] = df[columnas].round(1)
df.head()


customer_id churn  gender  senior_citizen partner dependents  tenure  \
0  0002_orfbo    no  female               0     yes        yes       9   
1  0003_mknfe    no    male               0      no         no       9   
2  0004_tlhlj   yes    male               0      no         no       4   
3  0011_igkff   yes    male               1     yes         no      13   
4  0013_exchz   yes  female               1     yes         no       3   

  phone_service multiple_lines internet_service  ... device_protection  \
0           yes             no              dsl  ...                no   
1           yes            yes              dsl  ...                no   
2           yes             no      fiber_optic  ...               yes   
3           yes             no      fiber_optic  ...               yes   
4           yes             no      fiber_optic  ...                no   

  tech_support streaming_tv streaming_movies        contract  \
0          yes          yes               no        one_year   
1           no           no              yes  month_to_month   
2           no           no               no  month_to_month   
3           no          yes              yes  month_to_month   
4          yes          yes               no  month_to_month   

  paperless_billing    payment_method daily_account charges_monthly  \
0               yes      mailed_check           2.2            65.6   
1                no      mailed_check           2.0            59.9   
2               yes  electronic_check           2.5            73.9   
3               yes  electronic_check           3.3            98.0   
4               yes      mailed_check           2.8            83.9   

   charges_total  
0          593.3  
1          542.4  
2          280.8  
3         1237.8  
4          267.4  

[5 rows x 22 columns]

In [143]:
# Verificar que la columna recien creada, 'daily_account' sea tipo float
df['daily_account'].dtype

dtype('float64')

In [144]:
# Traducir los encabezados de las columnas
df = df.rename(columns={
    'customer_id': 'id_cliente',
    'churn': 'desafiliado',
    'gender': 'genero',
    'senior_citizen': 'persona_mayor',
    'partner': 'pareja',
    'dependents': 'dependientes',
    'tenure': 'antiguedad',
    'phone_service': 'servicio_telefonico',
    'multiple_lines': 'multiples_lineas',
    'internet_service': 'servicio_de_internet',
    'online_security': 'seguridad_en_linea',
    'online_backup': 'copia_de_seguridad_en_linea',
    'device_protection': 'proteccion_de_dispositivos',
    'tech_support': 'soporte_tecnico',
    'streaming_tv': 'television_en_streaming',
    'streaming_movies': 'peliculas_en_streaming',
    'contract': 'contrato',
    'paperless_billing': 'facturacion_electronica',
    'payment_method': 'metodo_de_pago',
    'daily_account': 'cargo_diario',
    'charges_monthly': 'cargo_mensual',
    'charges_total': 'cargo_total'
})
df.head()

id_cliente desafiliado  genero  persona_mayor pareja dependientes  \
0  0002_orfbo          no  female              0    yes          yes   
1  0003_mknfe          no    male              0     no           no   
2  0004_tlhlj         yes    male              0     no           no   
3  0011_igkff         yes    male              1    yes           no   
4  0013_exchz         yes  female              1    yes           no   

   antiguedad servicio_telefonico multiples_lineas servicio_de_internet  ...  \
0           9                 yes               no                  dsl  ...   
1           9                 yes              yes                  dsl  ...   
2           4                 yes               no          fiber_optic  ...   
3          13                 yes               no          fiber_optic  ...   
4           3                 yes               no          fiber_optic  ...   

  proteccion_de_dispositivos soporte_tecnico television_en_streaming  \
0                         no             yes                     yes   
1                         no              no                      no   
2                        yes              no                      no   
3                        yes              no                     yes   
4                         no             yes                     yes   

  peliculas_en_streaming        contrato facturacion_electronica  \
0                     no        one_year                     yes   
1                    yes  month_to_month                      no   
2                     no  month_to_month                     yes   
3                    yes  month_to_month                     yes   
4                     no  month_to_month                     yes   

     metodo_de_pago cargo_diario cargo_mensual  cargo_total  
0      mailed_check          2.2          65.6        593.3  
1      mailed_check          2.0          59.9        542.4  
2  electronic_check          2.5          73.9        280.8  
3  electronic_check          3.3          98.0       1237.8  
4      mailed_check          2.8          83.9        267.4  

[5 rows x 22 columns]

In [145]:
# Traducir al español los registros

# Diccionario de traducción para todos los valores únicos
traduccion = {
    # Valores generales yes/no
    'yes': 'si',
    'no': 'no',

    # Género
    'female': 'femenino',
    'male': 'masculino',

    # Servicios específicos
    'no_phone_service': 'sin_servicio_telefonico',
    'no_internet_service': 'sin_servicio_de_internet',

    # Servicio de internet
    'dsl': 'dsl',
    'fiber_optic': 'fibra_optica',

    # Tipos de contrato
    'one_year': 'un_ano',
    'month_to_month': 'mes_a_mes',
    'two_year': 'dos_anos',

    # Métodos de pago
    'mailed_check': 'cheque_por_correo',
    'electronic_check': 'cheque_electronico',
    'credit_card_(automatic)': 'tarjeta_de_credito_(automatico)',
    'bank_transfer_(automatic)': 'transferencia_bancaria_(automatico)'
}

# Aplicar replace() a todo el DataFrame para traducir los valores
df = df.replace(traduccion)

df.tail()

id_cliente desafiliado     genero  persona_mayor pareja dependientes  \
7038  9987_lutyd          no   femenino              0     no           no   
7039  9992_rramn          si  masculino              0     si           no   
7040  9992_ujoel          no  masculino              0     no           no   
7041  9993_lhieb          no  masculino              0     si           si   
7042  9995_hotoh          no  masculino              0     si           si   

      antiguedad servicio_telefonico         multiples_lineas  \
7038          13                  si                       no   
7039          22                  si                       si   
7040           2                  si                       no   
7041          67                  si                       no   
7042          63                  no  sin_servicio_telefonico   

     servicio_de_internet  ... proteccion_de_dispositivos soporte_tecnico  \
7038                  dsl  ...                         no              si   
7039         fibra_optica  ...                         no              no   
7040                  dsl  ...                         no              no   
7041                  dsl  ...                         si              si   
7042                  dsl  ...                         si              no   

     television_en_streaming peliculas_en_streaming   contrato  \
7038                      no                     no     un_ano   
7039                      no                     si  mes_a_mes   
7040                      no                     no  mes_a_mes   
7041                      no                     si   dos_anos   
7042                      si                     si   dos_anos   

     facturacion_electronica      metodo_de_pago cargo_diario cargo_mensual  \
7038                      no   cheque_por_correo          1.8          55.2   
7039                      si  cheque_electronico          2.8          85.1   
7040                      si   cheque_por_correo          1.7          50.3   
7041                      no   cheque_por_correo          2.3          67.8   
7042                      no  cheque_electronico          2.0          59.0   

      cargo_total  
7038        742.9  
7039       1873.7  
7040         92.8  
7041       4627.6  
7042       3707.6  

[5 rows x 22 columns]

In [146]:
#Mostrar valores únicos por cada columna
for col in df.columns:
    print(f"\nColumna: '{col}'")
    print(df[col].unique())



Columna: 'id_cliente'
['0002_orfbo' '0003_mknfe' '0004_tlhlj' ... '9992_ujoel' '9993_lhieb'
 '9995_hotoh']

Columna: 'desafiliado'
['no' 'si']

Columna: 'genero'
['femenino' 'masculino']

Columna: 'persona_mayor'
[0 1]

Columna: 'pareja'
['si' 'no']

Columna: 'dependientes'
['si' 'no']

Columna: 'antiguedad'
[ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 37 49 66 67
 20 43 59 12 27  2 25 29 14 35 64 39 40 11  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 31 36 17 18 51 38 42
  0]

Columna: 'servicio_telefonico'
['si' 'no']

Columna: 'multiples_lineas'
['no' 'si' 'sin_servicio_telefonico']

Columna: 'servicio_de_internet'
['dsl' 'fibra_optica' 'no']

Columna: 'seguridad_en_linea'
['no' 'si' 'sin_servicio_de_internet']

Columna: 'copia_de_seguridad_en_linea'
['si' 'no' 'sin_servicio_de_internet']

Columna: 'proteccion_de_dispositivos'
['no' 'si' 'sin_servicio_de_internet']

Columna: 'soporte_tecnico'
['si' 'no' 'sin_servicio_de_internet']